crates/core/src/vm.rs

```diff
@@ -41,12 +41,12 @@ impl<'a> From<&'a Tx> for TxMode<'a> {
 
 //TODO: This is partially duplicated from the `vm` crate to avoid borrow checker issues
 //and pull all that crate in core. Will be revisited after trait refactor
-pub fn build_query<'a>(
-    ctx: &'a ExecutionContext,
+pub fn build_query<'a, I: 'a + Iterator<Item = RelValue<'a>>>(
+    ctx: &'a ExecutionContext<'a>,
     stdb: &'a RelationalDB,
-    tx: &'a TxMode,
+    tx: &'a TxMode<'a>,
     query: &'a QueryExpr,
-    sources: &mut SourceSet,
+    sources: &mut impl FnMut(SourceId) -> Option<I>,
 ) -> Result<Box<IterRows<'a>>, ErrorVm> {
```

Comment:
Needs comment describing the expected behavior of the `sources` function.

Reply:
I won't comment everywhere, but all the other functions that have been changed from `&mut SourceSet` to `&mut impl FnMut(SourceId) -> Option<I>` need the same comment, or at least a "see `build_query` for explanation of this argument."

Reply:
I'd also like this comment to strongly justify why these functions don't just take a `SourceSet` or `&mut SourceSet`, as I see this closure as a major increase in complexity and a barrier to at-a-glance understanding.